apps/condo/domains/miniapp/schema/preprocessors/B2BAppAccess.js

```diff
@@ -0,0 +1,505 @@
+ *         },
+ *         deletedAt, null,
+ *     },
+ *     deletedAt, null,
+ * }
+ */
+const getFilter = (pathToOrganizationId, organizationIds) => {
+    if (!isArray(pathToOrganizationId) || isEmpty(pathToOrganizationId)) throw new Error('"pathToOrganizationId" must be not empty array!')
+    if (!isArray(organizationIds)) throw new Error('"organizationId" must be array!')
+
+    if (pathToOrganizationId.length === 1) {
+        return {
+            [`${pathToOrganizationId[0]}_in`]: organizationIds,
+            deletedAt: null,
+        }
+    }
+
+    return {
+        [pathToOrganizationId[0]]: getFilter(pathToOrganizationId.slice(1), organizationIds),
+        deletedAt: null,
+    }
+}
+
+const canReadByServiceUser = async ({ authentication: { item: user }, args, listKey }, schemaConfig) => {
+    if (!user) return throwAuthenticationError()
+    if (user.deletedAt) return false
+
+    if (!listKey) return false
+
+    const pathToOrganizationId = get(schemaConfig, 'pathToOrganizationId', ['organization', 'id'])
+
+    const B2BAppContexts = await find('B2BAppContext', {
+        organization: { deletedAt: null },
+        app: {
+            accessRights_some: {
+                accessRightSet: {
+                    [`canRead${pluralize.plural(listKey)}`]: true,
+                    deletedAt: null,
+                },
+                user: { id: user.id, type: 'service', deletedAt: null },
+                deletedAt: null,
+            },
+            deletedAt: null,
+        },
+        status: 'Finished',
+        deletedAt: null,
+    })
+
+    const organizationIds = B2BAppContexts.map(ctx => ctx.organization)
+
+    if (!organizationIds || isEmpty(organizationIds)) return false
+
+    return getFilter(pathToOrganizationId, organizationIds)
+}
+
+const PERMISSION_FIELD = {
+    type: 'Checkbox',
+    defaultValue: false,
+    kmigratorOptions: { default: false },
+}
+
+const READ_ONLY_PERMISSION_FIELD = {
+    ...PERMISSION_FIELD,
+    access: {
+        read: true,
+        create: false,
+        update: false,
+    },
+}
+
+/**
+ *
+ * @param permissionFieldName {string}
+ * @return {string}
+ */
+const getSchemaDocForReadOnlyPermissionField = (permissionFieldName) => {
+    if (!isString(permissionFieldName) || permissionFieldName.trim().length < 1) throw new Error('"permissionFieldName" must be not empty string!')
+
+    if (permissionFieldName.startsWith('canRead')) {
+        return 'Currently, this field is read-only. You cannot get read access for the specified schema.'
+    }
+
+    if (permissionFieldName.startsWith('canManage')) {
+        return 'Currently, this field is read-only. You cannot get manage access for the specified schema.'
+    }
+
+    throw new Error('Permission field name no starts with "canManage" or "canRead"! You should check the implementation!')
+}
+
+const addPermissionFieldsToSchema = (schema, schemaName, readOnlyPermissionFields) => {
+    if (ALL_PERMISSION_NAMES.size < 1) {
```

Comment:
you use `ALL_PERMISSION_NAMES` before declaration

Reply:
Oh yeah, fixed

But this will not affect anything, because this function is called after const declaration